Fasterer

akofke opened the pull request:
Restructure PotentialMaster class to be faster by hardcoding it to work with certain types of potentials.  Most of the work is done within the PotentialMaster itself with callback functionality available for more detailed data.  The new structure is similar to that used within [etomica-cpp](https://github.com/etomica/etomica-cpp)

Remaining work includes:

- ~~Mean field long range correction~~
- ~~Compute (new, old) energy for one molecule~~
- ~~Extract neighbor list handling from PotentialMasterListFasterer into a separate class~~
- ~~Extract neighbor iteration into a separate class~~, eliminating need for separate PotentialMaster classes
- Implement callbacks for final results of computeAll as well as callbacks for ~~individual atom pairs~~
- ~~Develop structure to handle a collection of PotentialMaster-like classes, each handling a different type of potential contribution~~
- ~~Implement Ewald sum.  Real space is just a pair potential, but Fourier contribution is very different~~
- ~~Implement EAM, requires two loops over interacting neighbors~~
- ~~Handle non-spherical atomic potentials (dipole-dipole)~~
- Handle 3-body (ATM) potentials
- ~~Handle bond angle, torsion potentials~~
- ~~PotentialGroup is independent of new PotentialMaster class, but could be similarly streamlined~~
- Other molecular potentials ([P2Water4PSoft](https://github.com/etomica/etomica/blob/fasterer/etomica-core/src/main/java/etomica/models/water/P2Water4PSoft.java)) do not fit with the new classes.  We may decide to not handle them (require they be implemented via atomic contributions).
- ~~The pair PotentialMaster does not handle hard potential dynamics.  [IntegratorHard](https://github.com/etomica/etomica/blob/fasterer/etomica-core/src/main/java/etomica/integrator/IntegratorHard.java) might do more of the work here, but needs to know about neighboring atoms.~~
- ~~Implement 2nd derivative in potential (d2u)~~
- Try to port over other PotentialCalculations (need callbacks?)
- ~~grand canonical (update cell lists after insertion/removal of molecules)~~
- ~~NeighborCellManagerFasterer needs additional code to support slanty boxes~~

ajschult commented:
Handle bond angle, torsion potentials.  These are expressed as a function of a single angle.  The implementation can then be pretty simple.

angle:

void udu(double costheta, double[] u, double[] du);
du would be du/dcostheta.  PotentialMasterBonding would be responsible for computing costheta and to (if needed) compute the forces on each of the 3 atoms.  This is actually the same interface we have for Potential2Soft although the arguments mean different things.  Should we have another interface for such potentials or try to re-use?  some of the methods in Potential2Soft wouldn't make sense (long-range correction).

torsion and improper could follow the same approach.  

void udu(double cosphi, double[] u, double[] du);
du would be du/dcosphi. as described [elsewhere](http://cbio.bmt.tue.nl/pumma/index.php/Theory/Potentials), the PotentialMasterBonding need not treat torsion and improper angles differently.

ajschult commented:
PotentialGroup is independent of new PotentialMaster class, but could be similarly streamlined

Should this use the aggregate approach?  We need a molecular potential between molecules (pair, but we might have either spherical pair potentials or non-spherical, perhaps even 3-body) and within a molecule (both pair and bonding).

ajschult commented:
Handle non-spherical atomic potentials (dipole-dipole)

double u(Vector dr12, IAtom atom1, IAtom atom2);
double udu(Vector dr12, IAtom atom1, IAtom atom2, Vector g1, Vector g2);
double uduTorque(Vector dr12, IAtom atom1, IAtom atom2, Vector g1, Vector g2, Vector t1, Vector t2);

I'm tempted to drop the method that does gradients but not torques as this interface would be targeted at oriented atoms that would need torque.  The method that does torque could be udu.

The methods could have default implementations that call existing energy and gradientAndTorque.  We could stick the new methods in IPotentialTorque (or we could stick u in .  The new PotentialCompute would then be written similarly to PotnetialComputePair, but hold a 2D array of IPotentialTorque instead of Potential2Soft.

ajschult commented:
For the purpose of molecular potentials, a `DooDad` that holds the specification of bonding is more compelling to me, but it'd be bonding-only and I'd like to avoid requiring anyone to make it separately if they don't need to.  The `DooDad` could be how `PotentialMasterBonding` stores stuff, instead of holding it directly.  `isOnlyRigidMolecules`, `bondedAtoms`, `bondedPairs` and `bondedPartners` could be public members of the `DooDad`.  `PotentialMasterBonding` could take a `DooDad` at construction or it would make one if you call the existing constructor.  The `DooDad` would have `setBondingPotential`, `skipBondedPair` and `isOnlyRigidMolecules` (it would implement `BondingInfo`).  `PotentialMasterBonding `could still have its own `setBondingPotential`, but it would simply call `DooDad.setBondingPotential`.  And you could ask `PotentialMasterBonding` for the `DooDad` and hand it off to something else.  I can't imagine having both `PotentialMasterBonding` and a molecular potential in the same simulation; the main advantage is to not duplicate `setBondingPotential`, which will only get more complicated as we add 3- and 4-body bonding potentials.  We could drop `BondingInfo.makeBondingInfo` entirely; we could keep it, but it'd just call `pm.getBondingInfo()`.

ajschult commented:
The hard MD code I have here now runs fast for simple HS.  For square-well, it can do OK (about the same speed as master), but lacks the rigour we would want for research simulations (master has a special potential class to be rigourous).  It can handle square-well rigourously, but at greater computational cost (it needs to remember which pairs are within the well, currently in a hash map).  I can eliminate the cost by storing the state alongside the neighbor list, but I then need a special NeighborListManager that keeps track of it and a public field to expose it (to a separate iterator class, which calls back into a separate consumer).  And rinse and repeat for not-neighbor-listing using the hash map that the integrator currently uses.  And a healthy dose of casting.

ajschult committed:
does not need energy

ajschult committed:
add method to copy coordinates from another molecule/atom

ajschult committed:
implement iterAndSumAllNeighbors in NeighborManagerSimple

move SuperNbrConsumer into NeighborIterator

ajschult committed:
fix up PotentialComputePairGeneral to be not broken

ajschult committed:
fix up IAtom class that needs new copy method

ajschult committed:
add fasterer versions of ChainHSMD3D and GEMCWithRotation

ajschult committed:
implement new hard field interface

ajschult committed:
actually construct simple hard iterator

ajschult committed:
add fasterer versions of 2D simulations

ajschult committed:
add fasterer version of more simulation prototypes

ajschult committed:
don't need to be *that* noisy

ajschult committed:
use fij to compute virial, not cumulative fj

ajschult committed:
add flag so that pressure meter knows to ask for forces

ajschult committed:
add fasterer version of dimer MC simulation

ajschult committed:
Merge branch 'master' into fasterer

ajschult committed:
hard integrator takes potentials instead of computes

ajschult committed:
allow spherical potentials to be used in non-spherical compute

ajschult committed:
consistently consider z to be most important coordinate

ajschult committed:
add fasterer version of catalysis

add falseTime for decideBump so CO potential can find O-O vector

ajschult committed:
ensure our fieldPotentials aren't null before using them

ajschult committed:
handle getState properly for fixOverlap

ajschult committed:
avoid trying to update ourselves until we are set up

ajschult committed:
NeighborListManagerFastererHard will always need down neighbors

ajschult committed:
split the adding-a-neighbor part of checkNbrPair into a separate method

ajschult committed:
Add fasterer version of colloid module and lots of hard MD changes

use IAtom instead of IAtomKinetic for not-dynamic method
allow P1HardFieldGeneric to fix overlaps
allow P2HardFieldGeneric subclasses to provide alternate distances and energies for special atom pairs

ajschult committed:
PotentialComputeField needs to be a box listener too

ajschult committed:
add fasterer version of DCVGCMD module

ajschult committed:
don't print debugging anymore

ajschult committed:
expose our neighbor manager

ajschult committed:
add fasterer version of Droplet module

ajschult committed:
add a constructor that takes a Controller

ajschult committed:
add fasterer version of ensembles module

ajschult committed:
stop calling getAtoms when we don't need to

kofke committed:
Give user some indication when sampling of V and/or N is being constrained

ajschult committed:
stop pretending that our atomic potentials might have a torque

ajschult committed:
controller cleanup

ajschult committed:
skip computing old energy (integrator already has it)

ajschult committed:
keep KE up to date (it's cheap)

ajschult committed:
expose NeighborManager's BondingInfo

ajschult committed:
add a method that computes energy for multiple atoms (within the box)

ajschult committed:
allow a collision listener to be removed

ajschult committed:
fix hybrid MC; compute oldEnergy at the end of reset

handle alwaysScaleMomenta

ajschult committed:
don't write out timing unless we're supposed to

ajschult committed:
add computeManyAtomsOld, finish implementation for computeManyAtoms

ajschult committed:
add fasterer version of glass module

ajschult committed:
if a potential allows overlaps, we need to look for core escape

ajschult committed:
add collision time to the CollisionListener interface

ajschult committed:
add a (dirty) way to change potentials after integrator construction

ajschult committed:
add fasterer version of insertion module

ajschult committed:
implement iterAndSumAllNeighbors for neighbor listing

ajschult committed:
allow MeterPressureTensorFasterer to return just the virial tensor

ajschult committed:
remove stale debugging

ajschult committed:
misc cleanup

ajschult committed:
add fasterer version of InterfacialSW

ajschult committed:
split out zero-momentuming function as a separate class

ajschult committed:
add fasterer version of ljmd

ajschult committed:
pass callbacks to our sub-computes

ajschult committed:
add fasterer version of material fracture

kofke committed:
Remove debug code; make ghost atoms yellow

kofke committed:
Merge remote-tracking branch 'origin/fasterer' into fasterer

ajschult committed:
don't recurse

ajschult committed:
if j does not seem to be up from i, try down

ajschult committed:
add getBox to NeighborManager

handle 2D in cell manager

ajschult committed:
add falseTime to collisionTime method, deltaP to 1D bump method

ajschult committed:
implement iterAllNeighbors for neighbor-listing managers

ajschult committed:
add fasterer version of mu module

ajschult committed:
add fasterer versions of multiharmonic modules

ajschult committed:
add fasterer versions of SWVLE in nucleation

ajschult committed:
cleanup

ajschult committed:
add fasterer versions of osmosis

ajschult committed:
stop taking Simulation in PotentialCompute constructors

ajschult committed:
skip pairs with no potential

ajschult committed:
add falsetime to field collisionTime signature

ajschult committed:
replace Simulation with SpeciesManager in MoleculeAgentManager constructor

ajschult committed:
more avoiding passing Simulation around

ajschult committed:
don't pass Simulation to IntegratorHard

ajschult committed:
more passing SpeciesManager instead of Simulation

ajschult committed:
pass Box/IRandom instead of Simulation

ajschult committed:
don't pass Simulation to PotentialMasterMonatomic

ajschult committed:
add fasterer version of PistonCylinder

ajschult committed:
more avoid passing Simulation

ajschult committed:
pass IRandom to IntegratorMC instead of Simulation

ajschult committed:
drop Integrator constructors that take a Simulation

ajschult committed:
Avoid passing Simulation to MCMoves, PotentialMasters

ajschult committed:
more avoiding passing Simulation to constructors

ajschult committed:
final commit to avoid passing Simulation

ajschult committed:
species cannot be removed

ajschult committed:
if we're ignoring overlap, our energy is 0

ajschult committed:
potentials handle fixing overlap all by themselves now

ajschult committed:
we should reset the intergator after messing with the potential

ajschult committed:
Various changes for hard MD

Neighbor managers that fire events implement an interface
IntegratorHardFasterer handles null (periodic) collisions itself, needs to know diameters
create interface for hard neighbor iterators so methods can take falseTime
extract simple hard neighbor iterator as separate class

ajschult committed:
add fasterer version of ReactionEquilibrium

ajschult committed:
stop setting fixed random seeds

ajschult committed:
slow down the graphics

default mass => 12
repaint once per step
integrator time step => 0.05

ajschult committed:
add fasterer version of ReverseOsmosis (not water)

ajschult committed:
use centralImage instead of nearestImage (different for slanty boxes)

ajschult committed:
don't iterate over stale offsets

ajschult committed:
compute r du/dr properly, make qiqj mutable

ajschult committed:
make parameters into static fields

ajschult committed:
fix checks for array sizes (# of doubles, not # of complexes)

ajschult committed:
make switchFac a factory parameter

ajschult committed:
add SHAKE for fasterer as a listener

ajschult committed:
add fasterer version of reverse osmosis

ajschult committed:
don't fire an event for moving a deleted atom into a deleted atom's spot

ajschult committed:
allow bonding potential to be added more than once (replaces previous bond list)

in practice, only works if bonded atoms appear and disappear but are consistently bonded when present

ajschult committed:
add fasterer version of self assembly module

ajschult committed:
handle slanty boundaries in cell manager, add a test for cell manager

ajschult committed:
Merge remote-tracking branch 'origin/master' into fasterer

# Conflicts:
#	etomica-apps/src/main/java/etomica/dcvgcmd/DCVGCMD.java

ajschult committed:
fix NCM test: invoke init while atoms are in the box, don't print stuff

ajschult committed:
add fasterer version of statistics module

ajschult committed:
make nullCollisiontimes in setMaxCollisionDiameter in case other arrays have already been resized

ajschult committed:
don't try to compute collisions until we're ready

ajschult committed:
restart needs to happen RIGHT NOW; method is already called when controller is not active

ajschult committed:
add fasterer version of swmd

ajschult committed:
save our modifAction and bail if it changes

ajschult committed:
move truncation methods from Potential2Soft to IPotentialPair

ajschult committed:
handle LRC for non-spherical potentials

ajschult committed:
add fasterer version of VLE

kofke committed:
Generate list of files in etomica-apps before culling

Many files in apps will be removed to simplify migration to fasterer and ease subsequent development

This list of files may aid in identifying legacy classes that a future application could find useful

kofke committed:
Use vararg instead of array for constructor

ajschult committed:
add fasterer version of RATTLE

ajschult committed:
add fasterer version of RATTLE (the rest)

ajschult committed:
don't call into GUI thread from integrator thread

ajschult committed:
half-way through purging obsolete packages from apps

ajschult committed:
continue purging obsolete packages from apps

akofke committed:
Delete some unused classes that were broken from other removals

Remove main method from PotentialNonAdditive that was using
PotentialEmulCached

akofke committed:
Remove benchmark for VirialLJ which no longer exists

akofke committed:
Revive BenchSimVirialLJ and use VirialLJD

ajschult committed:
need to override u012add for FS

ajschult committed:
actually invoke preThermostat

velocities might be changed by preThermostat, so compute KE after

ajschult committed:
add listener that handles Nose-Hoover chains thermostat

ajschult committed:
finish purging old classes from etomica apps

ajschult committed:
remove exception thrown only when class needs to exist

ajschult committed:
remove a few more obsolete

ajschult committed:
migrate dcvgcmd (apps) to fasterer

ajschult committed:
give LJMD ability to use new NHC thermostat

ajschult committed:
just throw an exception

ajschult committed:
add fasterer version of liquidLJ (lattice sum still needs work)

ajschult committed:
clean up LRC integral method

ajschult committed:
use at least cellRange = 3

ajschult committed:
add fasterer version of mapped rdf

ajschult committed:
need to only loop to numCellOffsets since array may have garbage at the end

ajschult committed:
add fasterer version of mapped virial and energy

ajschult committed:
need to declare ourselves to be molecular for PotentialMasterLrc to understand

ajschult committed:
not setting an atom shouldn't disable molecular calculations

ajschult committed:
make non-spherical potentials implement Potential2Soft

throw exceptions if spherical methods called on actual non-spherical potential

ajschult committed:
don't reset agents from a non-existent agent manager

ajschult committed:
also compute torques (if appropriate) if forces are computed

ajschult committed:
clean up existing dielectric code

ajschult committed:
add fasterer versions of dipolar HS,LJ dieletric simulations

ajschult committed:
add callback capable of compute translation and rotation molecular Hessians

ajschult committed:
P2SoftSphericalSum can handle any potential

ajschult committed:
Add method that returns appropriate parameters for a given rCut

ajschult committed:
Need to update total Fourier energy after MC moves

ajschult committed:
Add fasterer version of dielectric simulation for TIP4P

ajschult committed:
Fix second derivative

ajschult committed:
Add callback for arbitrary Hessian contribution for a pair of atoms

ajschult committed:
Fix pair contribution, add callback for ewald

ajschult committed:
avoid doing multiplication twice

ajschult committed:
add a static method to impose PBC

ajschult committed:
add fasterer version of clathrate LD

ajschult committed:
actually do a mass-weighted sum

ajschult committed:
pull intramolecular correction out of loop over j molecules

ajschult committed:
include lattice sum images of a molecule with itself

ajschult committed:
handle Fourier intramolecular correction with PotentialMasterBonding

ajschult committed:
Force RR tensor to be symmetric

apparently correct, but doesn't match published equation

ajschult committed:
Merge branch 'master' into fasterer

# Conflicts:
#	etomica-apps/src/main/java/etomica/virial/simulations/SimulationVirial.java
#	etomica-apps/src/main/java/etomica/virial/simulations/hardsphere/VirialHSMixtureL.java

ajschult committed:
typeList needs an list for all types in the simulation

since we index by the global type index

ajschult committed:
r^-6 part needs a factor of 2

ajschult committed:
add a convenience method to return the global index of a molecule

ajschult committed:
don't store Hessian; just pass it off to something else

ajschult committed:
don't need these fields anymore

ajschult committed:
Add lattice dynamics class that handles wave vectors

ajschult committed:
Add convenience constructor that takes the atomic potentials

ajschult committed:
add a way for PotentialCallback to declare what atoms it wants

ajschult committed:
we don't need kCoefficients or wave vector field

ajschult committed:
actually compute cell-cell vectors so we get simple dynamical matrices

ajschult committed:
add a Ewald-Fourier subclass that is optimized for lattice dynamics

ajschult committed:
use new LD subclass for Ewald

ajschult committed:
just let people set the box and pretend it's helpful

ajschult committed:
add class for computing LJ+dipole+quad+induction+more

ajschult committed:
Merge remote-tracking branch 'origin/master' into fasterer

ajschult committed:
OK, it's non-periodic, but surely 10 is better than 1

ajschult committed:
snap step size to min/max

ajschult committed:
use COM by default

ajschult committed:
handle constraint map (for flexible molecules)

ajschult committed:
this line does nothing

ajschult committed:
include box index in exception

ajschult committed:
add a constructor that takes the pair potentials

ajschult committed:
add method to extract the array of potentials

ajschult committed:
pass pair potentials to constructor instead of method

ajschult committed:
compute cos3phi correctly (just do what udu does)

ajschult committed:
Add fasterer virial mcmoves, code in SimVirialOverlap2, alkane simulation

ajschult committed:
do fasterer

ajschult committed:
need to only mess with duAtom if we're doing new computation

if energy of the atoms is infinite, we can't subtract off any double-counted energy

ajschult committed:
add fasterer version of HSDimerNPT

ajschult committed:
add methods to take Potential2Soft array in constructor

ajschult committed:
add fasterer versions of HSNPT and HSNPTDimer

ajschult committed:
we don't do multi here, so no need to set all to 0

ajschult committed:
add neighbor-list version of computeOneInternal

ajschult committed:
add fasterer versions of LJ vacancy minimization, Frankel-Ladd integration sims

ajschult committed:
try to do a better job of converging iterations

ajschult committed:
if we're going to be added as a listener by default, there's gotta be some way to remove

ajschult committed:
gotta be this way, right?  right?!?!?!?

ajschult committed:
use better convergence for MuRoot too

ajschult committed:
add fasterer versions of HS vacancy sim

ajschult committed:
Split out vacancy color scheme, diameter hash, add LJ sim

ajschult committed:
fix up many-atoms compute

exclude intra-atom pairs appropriately during iteration instead of subtracting their contribution later
move like MCMoveAtomCoupledFasterer may need to avoid computeManyAtomsOld entirely due to potential - neighbor range mismatch and/or lattice sum

ajschult committed:
add fasterer version of LJ FCC HMA simulation

ajschult committed:
make BDAccFrac 1 by default (to match old behavior), but handle 0

ajschult committed:
add fasterer version of HMA for HCP LJ

ajschult committed:
add fasterer version of (HS) NPT demo

ajschult committed:
add fasterer version of Ashton-Wilding simulation for osmotic virial

ajschult committed:
handle vf=0 (gas-phase virial coefficient)

ajschult committed:
add fasterer version of restricted Gibbs ensemble osmotic virial coefficient

ajschult committed:
do no-drift for not-isothermal

ajschult committed:
actually compute minR2

ajschult committed:
actually exclude the atoms we're supposed to exclude

ajschult committed:
multiatomic molecules need a virial correction

ajschult committed:
fasterer version of rigid dynamics

ajschult committed:
don't always take velocity from molecule0

ajschult committed:
refactor CalcOrientation, add methods to get/set angular momentum

ajschult committed:
no no drift means drift

ajschult committed:
add more to CalcOrientation, add fasterer version of LJ dipole MD sim

ajschult committed:
default SpeciesManager as null doesn't actually work anymore

ajschult committed:
update coordinates for nodrift change (6e9d9cd)

uncomment TestSnapshots.main so it stays up to date

ajschult committed:
stop explicitly passing null for SpeciesManager

ajschult committed:
Handle bessel order=0 and order<0 for ints

ajschult committed:
Add fasterer version of Heisenberg

ajschult committed:
Add fasterer version of Ising app

ajschult committed:
set a default non-zero moment of inertia

ajschult committed:
implement many-atoms methods for bonding

ajschult committed:
implement MC and MD for star polymers

ajschult committed:
remove debug print

ajschult committed:
migrate virial simulations to fasterer

ajschult committed:
add file missing from fasterer virial commit

ajschult committed:
add methods to access fields

ajschult committed:
cleanup

ajschult committed:
allow builder to be returned

ajschult committed:
recompute collisions only if initialized, not if not initialized

ajschult committed:
add app to demonstrate strain

ajschult committed:
remove (almost) all classes obsoleted by fasterer classes

ajschult committed:
rename fasterer classes back to normal names

ajschult committed:
Merge remote-tracking branch 'origin/master' into fasterer

# Conflicts:
#	etomica-modules/src/main/java/etomica/modules/EtomicaModuleInfo.java

ajschult committed:
don't pass null for tracker

ajschult committed:
pass actual SpeciesManager instead of null

ajschult committed:
pass empty PotentialComputeAggregate instead of null

ajschult committed:
need to call NeighborListManager.init, including after volume change

ajschult committed:
code changed, update numerical values

ajschult committed:
update format due to death of "meta", update numerical values
